fv3core/stencils/a2b_ord4.py

```diff
@@ -22,485 +36,435 @@
-        g_in = dya[0, 1, 0] / dya
-        g_ou = dya[0, -2, 0] / dya[0, -1, 0]
-        qy[0, 0, 0] = 0.5 * (
-            ((2.0 + g_in) * qin - qin[0, 1, 0]) / (1.0 + g_in)
-            + ((2.0 + g_ou) * qin[0, -1, 0] - qin[0, -2, 0]) / (1.0 + g_ou)
-        )
 
+@gtscript.function
+def qy_edge_north2(qin: FloatField, dya: FloatFieldIJ, qy: FloatField):
+    g_in = dya[0, -1] / dya
+    return (
+        3.0 * (qin[0, -1, 0] + g_in * qin) - (g_in * qy[0, 1, 0] + qy[0, -1, 0])
+    ) / (2.0 + 2.0 * g_in)
 
-@gtstencil()
-def qy_edge_south2(qin: sd, dya: sd, qy: sd):
-    with computation(PARALLEL), interval(...):
-        g_in = dya / dya[0, -1, 0]
-        qy0 = qy
-        qy = (
-            3.0 * (g_in * qin[0, -1, 0] + qin) - (g_in * qy0[0, -1, 0] + qy0[0, 1, 0])
-        ) / (2.0 + 2.0 * g_in)
 
+@gtscript.function
+def great_circle_dist_noradius(p1a: Float, p1b: Float, p2a: Float, p2b: Float):
+    tb = sin((p1b - p2b) / 2.0) ** 2
+    ta = sin((p1a - p2a) / 2.0) ** 2
+    return asin(sqrt(tb + cos(p1b) * cos(p2b) * ta)) * 2.0
 
-@gtstencil()
-def qy_edge_north(qin: sd, dya: sd, qy: sd):
-    with computation(PARALLEL), interval(...):
-        g_in = dya[0, -2, 0] / dya[0, -1, 0]
-        g_ou = dya[0, 1, 0] / dya
-        qy[0, 0, 0] = 0.5 * (
-            ((2.0 + g_in) * qin[0, -1, 0] - qin[0, -2, 0]) / (1.0 + g_in)
-            + ((2.0 + g_ou) * qin - qin[0, 1, 0]) / (1.0 + g_ou)
-        )
 
+@gtscript.function
+def extrap_corner(
+    p0a: Float,
+    p0b: Float,
+    p1a: Float,
+    p1b: Float,
+    p2a: Float,
+    p2b: Float,
+    qa: Float,
+    qb: Float,
+):
+    x1 = great_circle_dist_noradius(p1a, p1b, p0a, p0b)
+    x2 = great_circle_dist_noradius(p2a, p2b, p0a, p0b)
+    return qa + x1 / (x2 - x1) * (qa - qb)
+
+
+def _a2b_ord4_stencil(
+    qin: FloatField,
+    qout: FloatField,
+    agrid1: FloatFieldIJ,
+    agrid2: FloatFieldIJ,
+    bgrid1: FloatFieldIJ,
+    bgrid2: FloatFieldIJ,
+    dxa: FloatFieldIJ,
+    dya: FloatFieldIJ,
+    edge_n: FloatFieldIJ,
+    edge_s: FloatFieldIJ,
+    edge_e: FloatFieldIJ,
+    edge_w: FloatFieldIJ,
+):
+    from __externals__ import REPLACE, i_end, i_start, j_end, j_start, namelist
 
-@gtstencil()
-def qy_edge_north2(qin: sd, dya: sd, qy: sd):
     with computation(PARALLEL), interval(...):
-        g_in = dya[0, -1, 0] / dya
-        qy0 = qy
-        qy = (
-            3.0 * (qin[0, -1, 0] + g_in * qin) - (g_in * qy0[0, 1, 0] + qy0[0, -1, 0])
-        ) / (2.0 + 2.0 * g_in)
-
-
-def ec1_offsets(corner):
-    i1a, i1b = ec1_offsets_dir(corner, "w")
-    j1a, j1b = ec1_offsets_dir(corner, "s")
-    return i1a, i1b, j1a, j1b
-
-
-def ec1_offsets_dir(corner, lower_direction):
-    if lower_direction in corner:
-        a = 0
-        b = 1
-    else:
-        a = -1
-        b = -2
-    return (a, b)
-
-
-def ec2_offsets_dirs(corner, lower_direction, other_direction):
-    if lower_direction in corner or other_direction in corner:
-        a = -1
```

Comment:
Wow! Great job unraveling all of these python functions!

Reply:
Thanks! It was a brain twister.